enterprise/coderd/prebuilds/metricscollector_test.go

```diff
@@ -307,6 +299,131 @@ func TestMetricsCollector(t *testing.T) {
+		value     *float64
+		isCounter bool
+	}
+
+	type testCase struct {
+		transition  database.WorkspaceTransition
+		jobStatus   database.ProvisionerJobStatus
+		initiatorID uuid.UUID
+		ownerID     uuid.UUID
+		metrics     []metricCheck
+		eligible    bool
+	}
+
+	test := testCase{
+		transition:  database.WorkspaceTransitionStart,
+		jobStatus:   database.ProvisionerJobStatusSucceeded,
+		initiatorID: agplprebuilds.SystemUserID,
+		ownerID:     agplprebuilds.SystemUserID,
+		metrics: []metricCheck{
+			{prebuilds.MetricCreatedCount, ptr.To(1.0), true},
+			{prebuilds.MetricClaimedCount, ptr.To(0.0), true},
+			{prebuilds.MetricFailedCount, ptr.To(0.0), true},
+			{prebuilds.MetricDesiredGauge, ptr.To(1.0), false},
+			{prebuilds.MetricRunningGauge, ptr.To(1.0), false},
+			{prebuilds.MetricEligibleGauge, ptr.To(1.0), false},
+		},
+		eligible: true,
+	}
+
+	logger := slogtest.Make(t, &slogtest.Options{IgnoreErrors: true})
+	clock := quartz.NewMock(t)
+	db, pubsub := dbtestutil.NewDB(t)
+	reconciler := prebuilds.NewStoreReconciler(db, pubsub, codersdk.PrebuildsConfig{}, logger, quartz.NewMock(t), prometheus.NewRegistry(), newNoopEnqueuer())
+	ctx := testutil.Context(t, testutil.WaitLong)
+
+	collector := prebuilds.NewMetricsCollector(db, logger, reconciler)
+	registry := prometheus.NewPedanticRegistry()
+	registry.Register(collector)
+
+	presetName := "default-preset"
+	defaultOrg := dbgen.Organization(t, db, database.Organization{})
+	setupTemplateWithDeps := func() database.Template {
+		template := setupTestDBTemplateWithinOrg(t, db, test.ownerID, false, "default-template", defaultOrg)
+		templateVersionID := setupTestDBTemplateVersion(ctx, t, clock, db, pubsub, defaultOrg.ID, test.ownerID, template.ID)
+		preset := setupTestDBPreset(t, db, templateVersionID, 1, "default-preset")
+		workspace, _ := setupTestDBWorkspace(
+			t, clock, db, pubsub,
+			test.transition, test.jobStatus, defaultOrg.ID, preset, template.ID, templateVersionID, test.initiatorID, test.ownerID,
+		)
+		setupTestDBWorkspaceAgent(t, db, workspace.ID, test.eligible)
+		return template
+	}
+
+	// When: starting with a regular template.
+	template := setupTemplateWithDeps()
+	labels := map[string]string{
+		"template_name":     template.Name,
+		"preset_name":       presetName,
+		"organization_name": defaultOrg.Name,
+	}
+
+	// nolint:gocritic // Authz context needed to retrieve state.
+	ctx = dbauthz.AsPrebuildsOrchestrator(ctx)
+
+	// Then: metrics collect successfully.
+	require.NoError(t, collector.UpdateState(ctx, testutil.WaitLong))
+	metricsFamilies, err := registry.Gather()
+	require.NoError(t, err)
+	require.NotEmpty(t, findAllMetricSeries(metricsFamilies, labels))
+
+	// When: the template is deleted.
+	require.NoError(t, db.UpdateTemplateDeletedByID(ctx, database.UpdateTemplateDeletedByIDParams{
+		ID:        template.ID,
+		Deleted:   true,
+		UpdatedAt: dbtime.Now(),
+	}))
+
+	// Then: metrics collect successfully but are empty because the template is deleted.
```

Comment:
I'm assuming this complies with monotonicity for counters because they're allowed to reset to 0 and that's essentially what this looks like to prometheus? So the new metric will have the same labelset and therefore be part of the same metric technically but it won't be a duplicate and the value will just represent the new template?

Reply:
Correct. Counters are allowed to reset.